src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java

```diff
@@ -419,36 +420,24 @@ private static DetailAST getFirstNonArrayDeclaratorParent(DetailAST ast) {
      * @return type node.
      */
     private static DetailAST getTypeLastNode(DetailAST ast) {
-        DetailAST result = ast.findFirstToken(TokenTypes.TYPE_ARGUMENTS);
-        if (result == null) {
-            result = getIdentLastToken(ast);
-            if (result == null) {
-                result = getArrayType(ast);
-            }
+        final DetailAST typeLastNode;
+        final DetailAST parent = ast.getParent();
+        final boolean isPrecededByTypeArgs =
+                parent.findFirstToken(TokenTypes.TYPE_ARGUMENTS) != null;
+        final Optional<DetailAST> objectArrayType = Optional.ofNullable(getIdentLastToken(ast));
+
+        if (isPrecededByTypeArgs) {
+            typeLastNode = parent.findFirstToken(TokenTypes.TYPE_ARGUMENTS)
+                    .findFirstToken(TokenTypes.GENERIC_END);
+        }
+        else if (objectArrayType.isPresent()) {
+            typeLastNode = objectArrayType.get();
```

Comment:
I tried to refactor this (get rid of the nested `if (result == null)`) in an initial `supplemental` commit, but it became too complicated and created a bunch of pitest failures, so this change had to be rolled into the main commit.